rust/common/cookieless/src/salt_cache.rs

```diff
@@ -0,0 +1,354 @@
+use base64::{engine::general_purpose, Engine};
+use std::sync::Arc;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+use crate::constants::{
+    MAX_NEGATIVE_TIMEZONE_HOURS, MAX_POSITIVE_TIMEZONE_HOURS, SALT_TTL_SECONDS,
+};
+use crate::metrics::metrics_consts::{
+    COOKIELESS_CACHE_HIT_COUNTER, COOKIELESS_CACHE_MISS_COUNTER, COOKIELESS_REDIS_ERROR_COUNTER,
+};
+use common_metrics::inc;
+use common_redis::{Client as RedisClient, CustomRedisError};
+use moka::sync::Cache;
+use rand::RngCore;
+use thiserror::Error;
+
+#[derive(Debug, Error, PartialEq)]
+pub enum SaltCacheError {
+    #[error("Date is out of range")]
+    DateOutOfRange,
+    #[error("Redis error: {0}")]
+    RedisError(String),
+    #[error("Failed to get salt from redis")]
+    SaltRetrievalFailed,
+}
+
+impl From<CustomRedisError> for SaltCacheError {
+    fn from(err: CustomRedisError) -> Self {
+        inc(
+            COOKIELESS_REDIS_ERROR_COUNTER,
+            &[("operation".to_string(), "from_redis_error".to_string())],
+            1,
+        );
+        SaltCacheError::RedisError(err.to_string())
+    }
+}
+
+/// SaltCache manages the local cache of salts used for cookieless hashing
+/// using the Moka synchronous caching library for efficient TTL handling
+pub struct SaltCache {
+    /// Cache of salts, keyed by YYYY-MM-DD
+    cache: Cache<String, Vec<u8>>,
+    /// Redis client for fetching and storing salts
+    redis_client: Arc<dyn RedisClient + Send + Sync>,
+}
+
+impl SaltCache {
+    /// Create a new SaltCache with the given Redis client
+    pub fn new(
+        redis_client: Arc<dyn RedisClient + Send + Sync>,
+        salt_ttl_seconds: Option<u64>,
+    ) -> Self {
+        // Create a cache with a maximum of 1000 entries
+        // This is more than enough for our use case, as we only store one salt per day
+        let cache = Cache::builder()
+            // Set TTL to the salt TTL
+            .time_to_live(Duration::from_secs(
+                salt_ttl_seconds.unwrap_or(SALT_TTL_SECONDS),
+            ))
+            // Build the cache
+            .build();
+
+        SaltCache {
+            cache,
+            redis_client,
+        }
+    }
+
+    /// Get the salt for a specific day (YYYY-MM-DD format)
+    pub async fn get_salt_for_day(&self, yyyymmdd: &str) -> Result<Vec<u8>, SaltCacheError> {
```

Comment:
This could be much faster if we stopped manipulating strings this often, but I understand the appeal of making this look just like the TS implementation. Reviewing it was a breeze!

Reply:
To be fair I would expect string manipulation to be pretty fast, at least compared to a redis round trip

When we have both the Rust and TS versions I'll look at doing some refactors to both

Reply:
Yeah, absolutely fair point. Time to reimplement this in Elixir and store it in a GenServer with shared memory across all our pods 🤣